src/method.c

```diff
@@ -215,6 +215,28 @@ static jl_value_t *resolve_globals(jl_value_t *expr, jl_module_t *module, jl_sve
             if (e->head == jl_method_sym || e->head == jl_module_sym || e->head == jl_throw_undef_if_not_sym) {
                 i++;
             }
+            if (e->head == jl_assign_sym && binding_effects) {
```

Comment:
Why is this allocating bindings if lowering told it not to (by omitting the `'(global ,lhs)` expr)?

Reply:
The intended design was that `'(global ,lhs)` affects scoping, but whether or not the presence of a GlobalRef allocates a mutable, global variable (as opposed to referencing an existing constant) depends on whether there's an assignment.

Reply:
That sounds somewhat breaking though, and a separate question to resolve later? As existing, we have it intended that `global` makes it a module global, regardless of what kind of assignment might occur later to it (so that you can potentially pre-reserve some names to prevent them from getting imported via `using` before you have a chance to assign them a value)
```
julia> f() = (global sin; sin)
f (generic function with 1 method)

julia> sin
ERROR: UndefVarError: `sin` not defined in `Main`
Suggestion: add an appropriate import or assignment. This global was declared but not assigned.
Hint: a global variable of this name also exists in Base.

julia> f() = cos
f (generic function with 2 methods)

julia> cos
cos (generic function with 14 methods)
```

Reply:
But also note that this IR construct `global` is separate and distinct from the AST construct `global`, and there is a pretty significant argument that we should give them different names. But whereas the later is a scope construct, the latter is toplevel instruction from lowering to allocate a global

Reply:
Yes, I want to change that, but that change is not in this PR. It's quite possible that this change is not required in this PR - I can try dropping it.